enhancements/insights/prometheus-alerts-for-insights-recommendations.md

```diff
@@ -0,0 +1,268 @@
+  remain hanging on the cluster indefinitely.
+- The alert should only fire if IO has been healthy at any point in the past 10
+  minutes. Healthy, in this context, means that there is no major error
+  happening (crashlooping, degraded operator status, etc.). A temporary error
+  state should not cause the alerts to be disabled.
+- If the user intentionally disables IO, the alerts should never fire. If IO is
+  disabled while some of the Insights alerts are firing, the alerts should
+  immediately stop firing (a part of their condition will be that IO is
+  enabled).
+- If IO was disabled by an accident, the user will be notified via the existing
+  `InsightsDisabled` alert, which is defined as a part of the Insights Operator.
+
+**Unhealthy Smart Proxy scenario:**
+
+- The latest successfully fetched and parsed report is too old
+  (`Meta.LastCheckedAt` is more than 8 hours ago)
+  - Reset the Insights recommendation metrics to avoid showing alerts about
+    recommendations that are no longer hitting.
+- Unable to fetch a report from Smart Proxy (endpoint unreachable, auth error, etc.)
+  - Log error, do not touch the Insights recommendation metrics.
+- The fetched report cannot be decoded into the expected structure, or the data
+  are otherwise invalid
+  - Log error, do not touch the Insights recommendation metrics.
+
+If there is anything wrong with the Smart Proxy or the reports returned by it,
+the metrics will not be modified until the validity of the last report expires.
+If IO is enabled, but something is wrong with the gathering or uploading, the
+metrics will not be directly affected for as long as the last report is valid.
+If IO has been disabled, or crashing for more then 10 minutes, the alerts will
+be disabled to avoid presenting the user with outdated information.
+
+### Risks and Mitigations
+
+- Disabling and re-enabling an alert (e.g, due to IO or Smart Proxy outage)
+  can trigger user-defined actions, even though there has not been any actual
+  change to the status of the recommendation. This could be very annoying if
+  it were to happen often or repeatedly over a short period of time.
+- The URL to detailed information about an active Insights recommendation will
+  be hard-coded into the alert template or the Insights Operator. If the
+  endpoint moves elsewhere, the URL will be broken and an updated version of the
+  alert template will have to be backported. The process can take several
+  months, and there is no guarantee that the user will perform the update.
+
+## Design Details
+
+### Open Questions [optional]
+
+1. Which operator conditions should fall under the unhealthy Insights Operator
+   state?
+1. How long should IO wait before considering a report to be no longer valid
+   (outdated)?
+1. If uploads keep failing, how and when will the alerts be disabled? (because
+   the current design assumes that the metrics are only modified after receiving
+   a report after a successful upload)
```

Comment:
I'd handle this via alerting on IO report freshness. Whenever there's no report fresher than 2-4x the reporting interval raise an info level alert. Once we cross over to 24hrs or something escalate to warning. But continue reporting metrics based on the most recently received report, is that stateful?

Does Insights have any latency SLOs? ie: condition on cluster happens at 00:00, how long does Insights have to expose that to the admin in either OCM or in-cluster?

Reply:
Okay, so if I understand correctly, the Insights Operator should never be in a Degraded state, because it never directly impacts the performance of the server.

If you believe it is a non-issue to keep the alerts enabled, and add an alert regarding the report "freshness", then I suppose it would be a reasonable option from the IO/implementation side as well. My only worry about this is that we will have to track several conditions separately (IO disabled, IO crashlooping, old report).

Regarding Insights SLOs, I will ask around to get a proper answer, but I believe there is _some_ interval within which an archive should typically be processed and a report become available for it. Typically, it happens in the order of minutes since its upload.

Reply:
I'm also concerned with latencies, especially with the "`Meta.LastCheckedAt` is more than 8 hours ago" mentioned below.  My impression is that worst case is:

1. Something goes wrong on the cluster.
2. Telemetry shipped up, so central Insights engine may already notice and have a report ready, but maybe not.
3. Insights operator collects Insights that include the bad thing.
4. Insights operator pushes that Insights tarball up to Red Hat.
5. Red Hat starts processing the new Insights tarball.
6. Insights operator asks for a recent Insights report.  Red Hat hasn't finished processing the upload from the previous step, so we get a stale report.
7. No Insights recommendations in the returned report.
8. Time passes (max 4h between upload attempts?).
9. Red Hat finishes processing the recent Insights tarball and produces recommendations.
10. New gather, new push, new report retrieved.
11. There are Insights recommendations in this new report!  Start serving metric time series for them, alerts go off.
12. Admin sees the alert and mitigates the issue.

And then that whole cycle again before the fact that the fix is landed makes it up off the cluster, gets analyzed, comes back down, and the alert clears.  Depending on caching in the recent-report-server, something like [`If-Modified-Since`][1] would make polling for a new report a bit cheaper, and maybe we could do that more frequently than we upload Insights tarballs themselves?

[1]: https://datatracker.ietf.org/doc/html/rfc7232#section-3.3

Reply:
Yes this is a good point. The key point is the number 6 IMO. I think we really need to check the time of the Insights report (and if it's older than let's say 15 min then do not update the metrics? or remove them?). Now the time information is part of the report content (HTTP body) and it would be probably better as some HTTP header, but I wouldn't poll the service (serving the reports) more often.

Reply:
I basically agree with what @tremes said here. The 8 hours were chosen rather arbitrarily and based on the previous feedback, I'm leaning towards the idea of keep the alerts enabled, and adding another metric/alert to indicate if we expected to receive a new report but do not have it yet (10 minutes since last successful upload, for example), which would stay on until we receive a report is within that time frame from its respective upload time.

The polling is something we wanted to avoid in order to prevent unnecessary load on the Smart Proxy. Instead, if the report fetched after a successful upload is older than expected, we will wait (e.g., with an exponential back-off) and try to get it later. We will keep trying until the next archive upload, but after some time (e.g., 10 minutes, as mentioned above) we would set the metric to indicate that the report is delayed.

Reply:
if you use the timestamp of the report (or the Insights upload that fed the report) as the metric value, you could have the alert say:

>`$DURATION` ago, there was an Insights recommendation `$SLUG`.  More details at `$URL`

that makes the stale-ness explicit, and my only concern is "users might think a stale alert was talking about their current cluster".  I'm ok with infinitely stale alerts if they explicitly declare how stale they are.

Reply:
Hmmm, so instead of having 0/1 as the metric values, we would use the timestamp from the latest report that contained the recommendation. When thinking about adding the timestamp to the metric, I was thinking of putting it into a label, but that seemed like a poor idea because it would change with every report. Putting it into the value instead makes a lot more sense, and if the Prometheus works the way you described it below, the recommendations that were not in the latest report would simply not have their respective metrics set. I'll have to discuss this with the team, and more importantly with someone who has a more experience with Prometheus to see if this could work for us.